src/binaryheap.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Datastructures: GAP package providing common datastructures.
+ * Licensed under the GPL 2 or later.
+ */
+
+#include "binaryheap.h"
+
+static UInt s_isLess_RNam = 0;  // FIXME: init these
+static UInt s_data_RNam = 0;      // FIXME: init these
+
+
+static void _BinaryHeap_BubbleUp_C(Obj data, Obj isLess, Int i, Obj elm)
+{
+    if ( LEN_PLIST( data ) < i ) {
+        GROW_PLIST( data, i );
+        SET_LEN_PLIST( data, i );
+        CHANGED_BAG( data );
+    }
+
+    while (i > 1) {
+        Obj parent = ELM_PLIST(data, i >> 1);
+        if (False == CALL_2ARGS(isLess, parent, elm))
+            break;
+        SET_ELM_PLIST( data, i, parent );
+        CHANGED_BAG( data );
+        i >>= 1;
+    }
+
+    SET_ELM_PLIST( data, i, elm );
+    CHANGED_BAG( data );
+}
+
+static Int _BinaryHeap_BubbleDown_C(Obj data, Obj isLess, Int i, Obj elm)
+{
+    Int len = LEN_PLIST(data);
+    while (2 * i <= len) {
+        Int left = 2 * i;
+        Int right = 2 * i + 1;
+        if (right > len || True == CALL_2ARGS(isLess, ELM_PLIST(data, right), ELM_PLIST(data, left))) {
```

Comment:
Another place to possibly use `LT`. Hm, I really want to rewrite this in C++ using templates